### PR TITLE
Allow custom filters in TissueMask

### PR DESCRIPTION
**Pull request by Christopher22**
## Description

This small patch allows a custom "pipeline" of image filters within the TissueMask to optimize for specific demands regarding tissue. The default behavior is not changed.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Fixes: #375 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/histolab/histolab/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:166]: This test smells to me. Here you are not testing the functionality. A unit test need to mock the objects involved in the actual implementation, here you are asserting something that does not imply nothing.

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
Thanks for your contribution @Christopher22 ❤️ !
I have some concerns on this PR 🤷🏽‍♂️ . The first thing is the tests you provided. Please read my comment. 
Furthermore @alessiamarcolini can give you other consideration on you work.

File: tests/unit/test_masks.py
Comment (Christopher22) [reply tests/unit/test_masks.py:166]: Thank you for the review! To be honest, I could not completely get your point... With the unit test above, I just check that the result after applying the custom pipeline differs from the default (I might do that more explicit, as I rely on the current documentation there...) Testing the filters themselves is already covered in the other tests, is it not?

Comment (ernestoarbitrio) [reply tests/unit/test_masks.py:166]: My point is that a unit tests that has to proof that your changes are working. So maybe I should've been more clear, sorry. Given that I'd like to see a unit tests that mocks your objects you introduced. For example how do we know that `_CustomFiltersComposition` has been called properly? Ho can we know that the `setter` within that is doing the right thing? Is it working only on a `Slide` or on a `Tile` too?

My hunch is that the changes of the custom filters works only if applied on a `Slide` and not on a `Tile` I'd like to see a tests with this kind of proof.

IMHO i do not think that the test `it_knows_custom_image_filters` is not applying the filters you believe. So to disprove my hypothesis I'd like to see a tests with this kind of assertions.

Hope I've been clear 🤷🏽‍♂️ . Let me know 😄

Comment (Christopher22) [reply tests/unit/test_masks.py:166]: You are of cause entirely right! I will go through the code again. The missing implementation for Tile is almost embarrassing ... 

Is there any documentation regarding your preferred unit tests and mock functions? They are amazingly sophisticated

Comment (ernestoarbitrio) [reply tests/unit/test_masks.py:166]: 👍🏽 thanks a lot. Anyway there's no official preference doc for unit testing. I'd suggest a quick reading of https://realpython.com/python-mock-library/. Furthermore if you want you can join the slack channel (link in the README) and we can discuss there... or even we can have a brief call and we can do the implementation together. Let me know what do you think

Comment (Christopher22) [reply tests/unit/test_masks.py]: Thank you very much for the additional information! I will join the Slack channel which seems like a valuable resource. 

I used my spare time on the train and added both the missing implementation and (hopefully) proper unit tests. 

What do you think?

File: histolab/filters/compositions.py
Comment (ernestoarbitrio) [inline histolab/filters/compositions.py:153]: Please could you add the missing return type and the type for the `custom_filters`

File: histolab/filters/compositions.py
Comment (ernestoarbitrio) [inline histolab/filters/compositions.py:149]: Please could you add the type for `custom_filters`?

File: histolab/filters/compositions.py
Comment (ernestoarbitrio) [inline histolab/filters/compositions.py:44]: Could you provide the missing typing annotation here?

File: histolab/filters/compositions.py
Comment (ernestoarbitrio) [inline histolab/filters/compositions.py:160]:
```suggestion

        Returns
```

File: histolab/filters/compositions.py
Comment (ernestoarbitrio) [inline histolab/filters/compositions.py:159]:
```suggestion
        """Filters composition for tissue estimation with custom filters.
```

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:140]: I would move this part of the docstring from the class and place it within the `__init__`. Furthermore I'd get rid of the last line `However ... `

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:25]:
```suggestion
from .filters.image_filters import ImageFilter, MorphologicalFilter
```

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:143]:
```suggestion
    def __init__(self, *filters: List[Union[ImageFilter, MorphologicalFilter]]) -> None:
```

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:146]: IMHO considering the https://www.python.org/dev/peps/pep-0257/ the first line of the docstring should be a sort of summary and fit in 88 chars. All the other info can be underneath the summary line with a black line between them.

Moreover the `Parameters` section of the docstring is missed 😄

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:192]:
```suggestion
        # Generate appropriate composition of filter
```

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:199]: Sorry but I see some problems in this portion of code:
1. `histolab.filters.image_filters.Compose` 👉🏽 could you import `Compose` without the whole reference of the histolab path?
2. The statement is too much convoluted maybe you should split it in something more readable (remember that the zen of python says: `beautiful is better than ugly 😂 `

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:223]: As before 😄 ... import `Compose` and avoid the usage of the whole path.

File: histolab/tile.py
Comment (ernestoarbitrio) [inline histolab/tile.py:171]: please keep the first line and the body separated with a blank line, and, as before, please follow the PEP 257 rules 🙏🏽 . Pay attention : The summary line must fit in one line (88chars)

File: histolab/tile.py
Comment (ernestoarbitrio) [inline histolab/tile.py:186]: same as before for the PEP.

Comment (ernestoarbitrio) [inline histolab/tile.py:186]:
```suggestion
        """Calculate the binary mask representing the tissue in the tile
        given ``filter_composition``.
```

File: tests/unit/test_compositions.py
Comment (ernestoarbitrio) [inline tests/unit/test_compositions.py:84]: IMHO this test is not meaningful. If you take a look at lines 52-58 you will see the assertions for the mocks. Could you please integrate those also in this test? Thanks

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:4]: please could avoid importing the whole module? I'd like to see the import statements only for the object needed 😄

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:148]:
```suggestion
```

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:180]:
```suggestion
```

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:193]:
```suggestion
```

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:196]:
```suggestion
```

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:159]: I'm not a huge fan of `assert_called_once`. The reason is that, yes, you're asserting that a proper method has been called, but the interesting info would be also know with which parameters? I'd suggest to switch to `assert_called_once_with` and check the calling args 😉

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:161]: same

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:188]: same

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:207]: same

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:209]: same

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
@Christopher22 I did this CR round with @alessiamarcolini.
Sorry for the bunch of request changes ... but overall looks neat.
Could you please address the comments in order to make us able to do a final round?
Thanks again for you effort 💪🏽 ... we really appreciated it 😄

**Comment by Christopher22** [comment]
Thank you so much for your comments! I hope, I was able to fix the things up you mentioned.

As a small refactor, I removed my 'set_custom_filters' and replaced it with a meaningful constructor which fits better in your existing codebase.

File: tests/unit/test_masks.py
Comment (Christopher22) [reply tests/unit/test_masks.py]: I try to remove all of the things. However, I sometimes have no access to the "self" parameter. In this case, I try to check only the remaining arguments.

Comment (ernestoarbitrio) [reply tests/unit/test_masks.py]: Sorry I do not understand this point. Could you elaborate? Thx

File: histolab/masks.py
Comment (ernestoarbitrio) [inline histolab/masks.py:197]: Could you import directly `Slide` and use it instead of the whole path?

File: tests/unit/test_masks.py
Comment (ernestoarbitrio) [inline tests/unit/test_masks.py:168]: i would remove this line. The reason is that in a test, for the sake of readability and for future developers we're trying to keep separated the 3 blocks that compose a test.
1. Objects declarations
2. Actual call
3. Assertions

Hope this can help

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
Hey @Christopher22 thanks for all your commitment. I left a couple of furhter minor comments. Could you please take a look? After that we are probably ready to go.

Heads up to @alessiamarcolini ☝🏽

**Comment by Christopher22** [comment]
Amazing! Your highlighed areas should be fixed by now :)

File: histolab/filters/compositions.py
Comment (alessiamarcolini) [inline histolab/filters/compositions.py:34]: docs: `*custom_filters parameter` is missing

File: histolab/filters/compositions.py
Comment (alessiamarcolini) [inline histolab/filters/compositions.py:46]:
```suggestion
    def __new__(cls, cls_,: type *custom_filters: imf.ImageFilter):
```

File: histolab/filters/compositions.py
Comment (alessiamarcolini) [inline histolab/filters/compositions.py:158]: as Codecov says, this line is not covered by tests (in this case it should be a unit test)

File: histolab/filters/compositions.py
Comment (alessiamarcolini) [inline histolab/filters/compositions.py:161]: same here, not covered by tests

File: histolab/filters/compositions.py
Comment (alessiamarcolini) [inline histolab/filters/compositions.py:174]: Is this possible? If yes, it needs a test

File: histolab/masks.py
Comment (alessiamarcolini) [inline histolab/masks.py:148]:
```suggestion
        *filters : ImageFilter
```
Otherwise I wouldn't know how to specify that that's an arbitrary number of parameters

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [inline tests/unit/test_masks.py:160]: not your responsibility, but since you touched this method... this line should also be removed

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [inline tests/unit/test_masks.py:175]: the problem of not checking the parameters can hide bugs in this very case: this test is not really checking that the mask is computed with the default filters. The method `Tile.calculate_tissue_mask` could be called:
1. by `Tile.tissue_mask`, with `FiltersComposition(Tile)` as parameter, if the code logic is correct OR
2. by _mask (dispatched to the `Tile`'s method `_`) with `custom_filters`, if the logic is not correct.

So please add the parameters in the assertion :)

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [reply tests/unit/test_masks.py:189]: Regarding the whole test: I feel that testing that `RgbToGrayscale` is called is not a good strategy, because:

1. usually we want to mock all the methods/instances that get called/used directly by the method we are testing, so in this case it would be `Tile.calculate_tissue_mask`
2. as a proof for my first point: this test is not really checking that the mask is built with the custom filters, because `RgbToGrayscale` is actually the first filter in the "default" filter sequence (for both `Slide` and Tile). So this test would actually pass with both the default and the "custom" filters represented by `[RgbToGrayscale()]` 🤓

Please let me know if this makes sense for you!

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
Apart from a couple of docstrings improvements, I had to reason carefully about two unit tests (see comments) and please try to cover all of your new code with tests (follow codecov comments), as we cannot merge this PR if it does not provide 100% coverage ✌🏼

**Comment by Christopher22** [comment]
Such cute apes! ;) Your comments should be integrated by now!

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [reply tests/unit/test_masks.py:192]: @Christopher22 I think you misunderstood my points here. This updated test covers point number 2 but not number 1, the mocked object is not directly called/used by the method you're testing.

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
Please check the comment I left about the `it_knows_its_mask_tile_and_supports_custom_filters` test.

PS please squash the commits at the end of the review, otherwise it's more difficult for us to be on track with the changes between review rounds ✌🏼

**Comment by Christopher22** [comment]
@alessiamarcolini Do you prefer a test such as this instead? :)

**Review by ernestoarbitrio — APPROVED** [review]

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [inline tests/unit/test_masks.py:186]: please use a np array here to be consistent with the real `calculate_tissue_mask` return type

File: tests/unit/test_masks.py
Comment (alessiamarcolini) [inline tests/unit/test_masks.py:206]: accordigly, here please use `np.testing.assert_array_equal`

**Review by alessiamarcolini — CHANGES_REQUESTED** [review]
Just a minor comment for consistency, the rest is perfect!! 👏🏼

**Comment by Christopher22** [comment]
Ready for squash ;)

**Review by alessiamarcolini — APPROVED** [review]
Let's merge it!

**Comment by alessiamarcolini** [comment]
@Christopher22 thank you so much for your valuable contribution! 👏🏼👏🏼

Now, to have a consistent code base, we would like to provide the same flexibility for the `BiggestTissueBoxMask` as well. Please let us know if you are interested in contributing again in this regard! Otherwise, we can take care of it, no pressure at all 😉

**Pull request merged**
